dtm0/recovery.c

```diff
@@ -0,0 +1,2170 @@
+}
+
+static int recovery_machine_log_iter_init(struct m0_dtm0_recovery_machine *m,
+					  struct m0_be_dtm0_log_iter      *iter)
+{
+	M0_PRE(m->rm_ops.log_iter_init != NULL);
+	return m->rm_ops.log_iter_init(m, iter);
+
+}
+
+static void recovery_machine_log_iter_fini(struct m0_dtm0_recovery_machine *m,
+					   struct m0_be_dtm0_log_iter      *iter)
+{
+	M0_PRE(m->rm_ops.log_iter_fini != NULL);
+	m->rm_ops.log_iter_fini(m, iter);
+}
+
+static void recovery_machine_redo_post(struct m0_dtm0_recovery_machine *m,
+				       struct m0_fom                   *fom,
+				       const struct m0_fid *tgt_proc,
+				       const struct m0_fid *tgt_svc,
+				       struct dtm0_req_fop *redo,
+				       struct m0_be_op *op)
+{
+	M0_PRE(m->rm_ops.redo_post != NULL);
+	m->rm_ops.redo_post(m, fom, tgt_proc, tgt_svc, redo, op);
+}
+
+static void recovery_machine_recovered(struct m0_dtm0_recovery_machine *m,
+				       const struct m0_fid *tgt_proc,
+				       const struct m0_fid *tgt_svc)
+{
+	M0_PRE(m->rm_ops.ha_event_post != NULL);
+	m->rm_ops.ha_event_post(m, tgt_proc, tgt_svc,
+				M0_CONF_HA_PROCESS_DTM_RECOVERED);
+}
+
+static void recovery_machine_lock(struct m0_dtm0_recovery_machine *m)
+{
+	m0_sm_group_lock(&m->rm_sm_group);
+}
+
+static void recovery_machine_unlock(struct m0_dtm0_recovery_machine *m)
+{
+	m0_sm_group_unlock(&m->rm_sm_group);
+}
+
+enum recovery_fom_state {
+	RFS_INIT = M0_FOM_PHASE_INIT,
+	RFS_DONE = M0_FOM_PHASE_FINISH,
+	RFS_WAITING,
+	RFS_FAILED,
+	RFS_NR,
+};
+
+static struct m0_sm_state_descr recovery_fom_states[] = {
+	[RFS_INIT] = {
+		.sd_name      = "RFS_INIT",
+		.sd_allowed   = M0_BITS(RFS_WAITING, RFS_DONE, RFS_FAILED),
+		.sd_flags     = M0_SDF_INITIAL,
+	},
+	/* terminal states */
+	[RFS_DONE] = {
+		.sd_name      = "RFS_DONE",
+		.sd_allowed   = 0,
+		.sd_flags     = M0_SDF_TERMINAL,
+	},
+	/* failure states */
+	[RFS_FAILED] = {
+		.sd_name      = "RFS_FAILED",
+		.sd_allowed   = M0_BITS(RFS_DONE),
+		.sd_flags     = M0_SDF_FAILURE,
+	},
+
+	/* intermediate states */
+#define _ST(name, allowed)            \
+	[name] = {                    \
+		.sd_name    = #name,  \
+		.sd_allowed = allowed \
+	}
+	_ST(RFS_WAITING,           M0_BITS(RFS_DONE,
+					   RFS_FAILED, RFS_WAITING)),
+#undef _ST
+};
+
+const static struct m0_sm_conf recovery_fom_conf = {
+	.scf_name      = "recovery_fom",
+	.scf_nr_states = ARRAY_SIZE(recovery_fom_states),
+	.scf_state     = recovery_fom_states,
+};
+
+static size_t recovery_fom_locality(const struct m0_fom *fom)
+{
+	return 1;
+}
+
+static const struct m0_fom_ops recovery_fom_ops = {
+	.fo_fini          = recovery_fom_self_fini,
+	.fo_tick          = recovery_fom_tick,
+	.fo_home_locality = recovery_fom_locality
+};
+
+static struct m0_fom_type recovery_fom_type;
+static const struct m0_fom_type_ops recovery_fom_type_ops = {};
+
+M0_INTERNAL int m0_drm_domain_init(void)
+{
+	int         rc = 0;
+
+	if (!m0_sm_conf_is_initialized(&m0_drm_sm_conf)) {
```

Comment:
We should M0_ASSERT(!m0_sm_conf_is_initialized(&m0_drm_sm_conf)) here.

If this is not satisfied, something must be wrong there - multiple calling of this m0_drm_domain_init() is a bug.

Reply:
3:Working on it.

Reply:
Fixed/merged.